### PR TITLE
Routing / Fetching data

### PR DESCRIPTION
**Pull request by aravind289**

### CODE REVIEW

File: src/components/fetch.jsx [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+      data: [],
+    }
+  }
+
+  componentDidMount() {
+    fetch('https://jsonplaceholder.typicode.com/users')
+      .then(res => res.json())
+      .then(output => {
+        this.setState({ data: output })
+        console.log(this.state.data)
```
[the remark this answers is not in the document]
Comment (aravind289) [reply src/components/fetch.jsx]: ok

File: src/components/category.jsx [hunk 2]
```diff
@@ -35,15 +36,11 @@ class Fetch extends React.Component {
...
+    const { ctg } = this.state
     return (
       <div>
-        {this.state.data.map(item => (
-          <h4 key={item._id}>{item.name}</h4>
-        ))}
-        <div>
-          <p>bravo </p>
-        </div>
+        <h2>{ctg._id}</h2>
```
Comment (digi0ps) [inline src/components/category.jsx]: Don't have to print id. Remove this line.

File: src/components/category.jsx [hunk 3]
```diff
@@ -5,16 +5,17 @@ class Fetch extends React.Component {
   constructor() {
     super()
     this.state = {
-      data: [],
+      ctg: [],
```
Comment (digi0ps) [inline src/components/category.jsx]: Name it `category`. It shouldn't be a reserved word.

File: src/components/category.jsx [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+      .then(output => {
+        this.setState({ ctg: output, isloading: false })
+        console.log(output)
+      })
+  }
+
+  render() {
+    if (this.state.isloading) {
+      return (
+        <div>
```
Comment (digi0ps) [inline src/components/category.jsx]: Make a separate `Loader` component, import it and then use it here and everywhere else.

File: src/components/category.jsx [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+          <p>loading....</p>
+          <img src={Loader} />
+        </div>
+      )
+    }
+    if (this.state.error) {
+      retur(
+        <div>
+          <p>something wrong!!!!!</p>
```
Comment (digi0ps) [inline src/components/category.jsx]: `return` and similarly make a separate `Error` component for further use.

Comment (digi0ps) [reply src/components/category.jsx]: This two pieces of code is just repeated everywhere, so when you make a separate component you can reuse it everywhere.

File: .vscode/settings.json [hunk 6]
```diff
@@ -1,5 +1,5 @@
 {
-  "workbench.colorTheme": "Visual Studio Light",
+  "workbench.colorTheme": "Default Dark+",
   "editor.renderWhitespace": "all",
```
Comment (digi0ps) [inline .vscode/settings.json]: Remove this file.

File: src/components/category.jsx [hunk 7]
```diff
@@ -0,0 +1,48 @@
+import React from 'react'
+import Loader from './loader'
+import Error from './error'
```
Comment (digi0ps) [inline src/components/category.jsx]: Don't name it `Error`, it's a reserved keyword. Might cause probs. So change this to something like "ErrorComponent".

File: src/components/category.jsx [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+      .then(output => {
+        this.setState({ category: output, isloading: false })
+        console.log(output)
+      })
+  }
+
+  render() {
+    if (this.state.isloading) {
+      return (
+        <div>
```
Comment (digi0ps) [inline src/components/category.jsx]: There is no need for this <div> wrapping the loader. Just `return <Loader />`.

File: src/components/category.jsx [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+          <Loader />
+        </div>
+      )
+    }
+    if (this.state.error) {
+      retur(
+        <div>
```
Comment (digi0ps) [inline src/components/category.jsx]: Same, remove the wrapping div. And it's `return`.

File: src/components/package.jsx [hunk 10]
```diff
@@ -0,0 +1,81 @@
+import React from 'react'
+import Loader from './loader'
+import Image from '../images/logo-visit.png'
+import Category from './category'
```
Comment (digi0ps) [inline src/components/package.jsx]: We will be changing the file structure.

File: src/components/package.jsx [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+import Package from './package_id'
+import { Switch, BrowserRouter as Router, Route, Link } from 'react-router-dom'
+import Error from './error'
+
+class Fetch extends React.Component {
+  constructor() {
+    super()
+    this.state = {
+      pkg: [],
```
Comment (digi0ps) [inline src/components/package.jsx]: `packages`, `categories`

File: src/components/package.jsx [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+      .then(pkg_output => {
+        this.setState({ pkg: pkg_output, isloading: false })
+        console.log(pkg_output)
+      })
+      .catch(error => this.setState({ error, isloading: false }))
+
+    fetch('http://127.0.0.1:3000/api/v1/category/all')
+      .then(res => res.json())
+      .then(category_output => {
+        this.setState({ category: category_output, isloading: false })
```
Comment (digi0ps) [inline src/components/package.jsx]: **IMPORTANT**

Not the right way to execute two fetch requests.
First call package url, and then once you got the output in `then`, call fetch category url inside it. Once you have both categories and packages, then set all the states.

File: src/components/router.jsx [hunk 13]
```diff
@@ -0,0 +1,23 @@
+import React from 'react'
+import { Switch, BrowserRouter as Router, Route, Link } from 'react-router-dom'
```
Comment (digi0ps) [inline src/components/router.jsx]: Importing Link is not needed here, remove it.

File: src/components/home/index.js [hunk 14]
```diff
@@ -0,0 +1,8 @@
+import React from 'react'
+import Home from './home'
+
+export default class Home extends React.Component {
+  render() {
+    return <Home />
+  }
+}
```
Comment (digi0ps) [inline src/components/home/index.js]: You are complicating things.
Remove react import.
and then just export home.
```
import Home from './home'

export default Home
```
That's it.

File: src/components/package_id/index.js [hunk 15]
```diff
@@ -0,0 +1,8 @@
+import React from 'react'
+import Package from './package_id'
+
+export default class Package_id extends React.Component {
```
Comment (digi0ps) [inline src/components/package_id/index.js]: Folder name should be `package` and not `package_id`.

File: src/components/package_id/index.js [hunk 16]
```diff
@@ -0,0 +1,8 @@
...
+  render() {
+    return <Package />
```
Comment (digi0ps) [inline src/components/package_id/index.js]: And export like I said for home.

File: src/components/category_id/category_id.jsx [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+
+class Category extends React.Component {
+  constructor() {
+    super()
+    this.state = {
+      category: [],
+      isloading: true,
+    }
+  }
+
```
Comment (digi0ps) [inline src/components/category_id/category_id.jsx]: Folder name should be `category`.

**Pull request merged**